### DISCUSSION TITLE
What is CorrectableCodeSmell?

### DISCUSSION THREAD

**Discussion by pinkasey** [detekt/detekt#3933] [Q&A] [answered]
I've tries to understand what is `CorrectableCodeSmell` for, and how to use it in my custom rules.
It sounds useful - some of the custom-rules I wrote can auto-correct the smell they find easily.

The only attribute `CorrectableCodeSmell` adds on top of `CodeSmell` is `autoCorrectEnabled: Boolean` -
How do I tell Detekt **how** to auto-correct the smell?

I didn't find help in the documentation, or the code, or by googling `CorrectableCodeSmell`.

**Comment by cortinico** [comment, marked as answer]
`CorrectableCodeSmell` is defined here: https://github.com/detekt/detekt/blob/5048b8bf0bb6a64de80f3d0b5aa25f694d71167d/detekt-api/src/main/kotlin/io/gitlab/arturbosch/detekt/api/CodeSmell.kt#L44-L74

If you're planning to use the `autoCorrect` feature in your custom rule, you should make sure your rule returns a `CorrectableCodeSmell` instance. From the practical point of view, a `CorrectableCodeSmell` is really similar to a classic `CodeSmell`, modulo a flag that tells if the report is correctable or not.

To implement a rule with auto-correction capability, you should check if the field `autoCorrect` is set to `true` in your rule. If so, you can edit the AST as you wish to correct the reporting.

You can see an example of it in #3811

**Reply by FallenDemon2000** [reply]
Although it has been quite some time, just passing to appreciate the answer :raised_hands: 
As someone who's just trying to figure this out, this was really helpful :heart:
